### PR TITLE
Add callbacks

### PR DESCRIPTION
**Pull request by toddkao**
Add callbacks

- onWalletConnected
- onWalletDisconnected
- onTransactionBroadcasted
- onTransactionComplete
- onTransactionFailed

### CODE REVIEW

File: packages/widget/src/hooks/useCreateCosmosWallets.tsx [hunk 1]
```diff
@@ -29,12 +29,14 @@ import { useCallback } from "react";
 import { skipAssetsAtom, skipChainsAtom } from "@/state/skipClient";
 import { sourceAssetAtom } from "@/state/swapPage";
 import { isMobile } from "@/utils/os";
+import { allCallbacksAtom } from "@/state/callbacks";
```
Comment (ericHgorski) [inline packages/widget/src/hooks/useCreateCosmosWallets.tsx:32]: In terms of naming, what do you think about just `callbacks` or `walletCallbacks` ? (for this atom as well as rest of references)

File: packages/widget/src/hooks/useCreateCosmosWallets.tsx [hunk 2]
```diff
@@ -101,6 +103,7 @@ export const useCreateCosmosWallets = () => {
           },
           disconnect: async () => {
             console.error("Prax wallet is not supported");
+            allCallbacks?.onWalletDisconnected?.({ chainType: "cosmos" });
```
Comment (ericHgorski) [inline packages/widget/src/hooks/useCreateCosmosWallets.tsx:106]: use `ChainType` enum from client lib

File: packages/widget/src/hooks/useCreateCosmosWallets.tsx [hunk 3]
```diff
@@ -260,6 +286,7 @@ export const useCreateCosmosWallets = () => {
           disconnect: async () => {
             await disconnectAsync();
             setCosmosWallet(undefined);
+            allCallbacks?.onWalletDisconnected?.({ chainType: "cosmos" });
```
Comment (ericHgorski) [inline packages/widget/src/hooks/useCreateCosmosWallets.tsx:289]: use `ChainType` enum from client lib

File: packages/widget/src/hooks/useCreateSolanaWallets.tsx [hunk 4]
```diff
@@ -71,14 +86,14 @@ export const useCreateSolanaWallets = () => {
         disconnect: async () => {
           await wallet.disconnect();
           setSvmWallet(undefined);
-          // TODO: onWalletDisconnected
+          allCallbacks?.onWalletDisconnected?.({ chainType: "svm" });
```
Comment (ericHgorski) [inline packages/widget/src/hooks/useCreateSolanaWallets.tsx:89]: use `ChainType` enum from client lib

File: packages/widget/src/hooks/useCreateCosmosWallets.tsx (hunk 1 above)
Comment (codingki) [reply packages/widget/src/hooks/useCreateCosmosWallets.tsx:32]: agree on this one

File: packages/widget/src/state/callbacks.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  txHash: string;
+  chainId: string;
+  explorerLink: string;
+};
+
+export type OnTransactionFailedProps = {
+  error: string;
+};
+
+export type AllCallbacks = {
```
Comment (codingki) [inline packages/widget/src/state/callbacks.ts:36]: rename to Callbacks

**Comment by ericHgorski** [comment]
Im not sure we should pass chainType to on wallet disconnect, as this gets confusing with multi-chain wallets (e.g. Keplr currently gets chainType: evm)

could we pass out wallet name instead?

**Comment by ericHgorski** [comment]
Also for on wallet connectedI think its fine if we just pass `chainIdToAddressMap` and not `chainIds` as they can derive `chainIds` easily from the map. what do you think?

**Comment by ericHgorski** [comment]
All the transaction stuff worked great!

**Comment by toddkao** [comment]
> Im not sure we should pass chainType to on wallet disconnect, as this gets confusing with multi-chain wallets (e.g. Keplr currently gets chainType: evm)
> 
> could we pass out wallet name of disconnected wallet instead?

I think passing wallet name when disconnecting makes sense, but I also think specifying chainType also makes sense, as you said, you could have Keplr connected for both cosmos and EVM, when you disconnect one of them I think the idea is that if we didn't return chainType you wouldn't know if you disconnected Keplr for cosmos chains or for EVM chains

**Comment by ericHgorski** [comment]
> > Im not sure we should pass chainType to on wallet disconnect, as this gets confusing with multi-chain wallets (e.g. Keplr currently gets chainType: evm)
> > could we pass out wallet name of disconnected wallet instead?
> 
> I think passing wallet name when disconnecting makes sense, but I also think specifying chainType makes sense, as you said, you could have Keplr connected for both cosmos and EVM, when you disconnect one of them I think the idea is that if we didn't return chainType you wouldn't know if you disconnected Keplr for cosmos chains or for EVM chains

Sounds good im aligned

**Comment by ericHgorski** [comment]
> > Also for on wallet connectedI think its fine if we just pass `chainIdToAddressMap` and not `chainIds` as they can derive `chainIds` easily from the map. what do you think?
> 
> I don't think it's a big deal to leave it as is, but I don't feel too strongly about it

Ok lets remove it then! If I saw both of those it would make me think they were derived differently and I would feel confused. 

The only last thing that would be a nice to have is destructuring the props in the next-js example so if I was developer reading it, I could easily tell what info args I get in the callback instead of `props` in general. Otherwise LGTM! thanks for this

**Review by ericHgorski — APPROVED** [review]
LGTM!

**Pull request merged**